### PR TITLE
URI Change to GUID; Tests for IObjects and BHoMObjects (PR #41)

### LINKED ISSUE #41: GUID get changed anytime we re-create identical objects, which can create inconsistencies with the graph data
#### Description:

In order to compare design options (or parts of the changes of the graph), it is important to be consistent with the GUIDs and URIs in the graph. However, following BHoM conventions, anytime we recompute the grasshopper definition, new GUIDs get assigned to objects, followingly also new URIs to nodes. Is there a way to be consistent with object GUIDs? I wonder how you deal with this in general when working with BHoM objects.  Or do you query the object one it is created, and not initiate it anymore? 

SensorData

### PR DESCRIPTION
**Pull request by polnischfrosch**
### Issues addressed by this PR

Closes #41 

Changed the composition of Individuals URI from Hash to GUID
Created new Tests in ToTTLTests to check the new URI for IObjects, BHoMObjects and non serialized IGeometry Objects.

### CODE REVIEW

File: RDF_Tests/ToTTLTests.cs [hunk 1]
```diff
@@ -384,6 +388,58 @@ public static void NurbsCurve_ControlPoints()
...
+        {
+            Room room = new Room();
+            room.Perimeter = new Polyline() { ControlPoints = new List<Point>() { new Point(), new Point() { X = 5 }, new Point() { X = 99 } } };
+            room.Location = new Point();
+            room.Name = "first room object";
+
+            CSharpGraph cSharpGraph = Compute.CSharpGraph(new List<object>() { room }, m_ontologySettings);
+            string TTLGraph = cSharpGraph.ToTTL();
+
+            Assert.IsTrue(TTLGraph.Contains(room.BHoM_Guid.ToString()));
```
Comment (alelom) [inline RDF_Tests/ToTTLTests.cs:402]: This check should be replaced with:
```cs
string roomUrl = Flurl.Url.Combine(cSharpGraph.OntologySettings.ABoxSettings.IndividualsBaseAddress, room.BHoM_Guid.ToString());
Assert.IsTrue(TTLGraph.Contains(roomUrl));
```
(will add a commit for you)

File: RDF_Tests/ToTTLTests.cs [hunk 2]
```diff
@@ -384,6 +388,58 @@ public static void NurbsCurve_ControlPoints()
...
+            CSharpGraph cSharpGraph = Compute.CSharpGraph(new List<object>() { nurbs }, new OntologySettings()
+            {
+                TBoxSettings = new TBoxSettings { GeometryAsOntologyClass = true }
+            });
+            string TTLGraph = cSharpGraph.ToTTL();
+            IObject iObject = nurbs as IObject;
+            string hash = BH.Engine.Base.Query.Hash(iObject);
+            string nurbsGuid = Engine.Adapters.RDF.Query.GuidFromString(hash).ToString();
+
+            Assert.IsTrue(TTLGraph.Contains(nurbsGuid));
```
Comment (alelom) [inline RDF_Tests/ToTTLTests.cs:434]: This check should be replaced with:
```cs
string nurbsUrl = Flurl.Url.Combine(cSharpGraph.OntologySettings.ABoxSettings.IndividualsBaseAddress, nurbsGuid.ToString());
Assert.IsTrue(TTLGraph.Contains(nurbsUrl));
```
(will add a commit for you)

**Review by alelom — CHANGES_REQUESTED** [review]
Thanks a lot for this @polnischfrosch ! Looks great and I really liked the reasoning you've put.
However, there probably is still some work to be done. If you inspect the output TTL of the first test case you've made (the one about the `Room`) you will see, under individuals:
```
### http://individuals.address/10b49b3d-f976-bb2a-4a04-2a7dc83a56b8
<http://individuals.address/10b49b3d-f976-bb2a-4a04-2a7dc83a56b8> rdf:type owl:NamedIndividual ,
		:BH.oM.Architecture.Elements.Room ;
		:BH.oM.Architecture.Elements.Room.Perimeter "ew0KICAiJHR5cGUiOiAiQkguRW5naW5lLkFkYXB0ZXJzLlJERi5Db252ZXJ0K1R5cGVXcmFwcGVyYDFbW0JILm9NLkdlb21ldHJ5LlBvbHlsaW5lLCBHZW9tZXRyeV9vTV1dLCBSREZfRW5naW5lIiwNCiAgIlZhbHVlIjogew0KICAgICIkdHlwZSI6ICJCSC5vTS5HZW9tZXRyeS5Qb2x5bGluZSwgR2VvbWV0cnlfb00iLA0KICAgICJDb250cm9sUG9pbnRzIjogew0KICAgICAgIiR0eXBlIjogIlN5c3RlbS5Db2xsZWN0aW9ucy5HZW5lcmljLkxpc3RgMVtbQkgub00uR2VvbWV0cnkuUG9pbnQsIEdlb21ldHJ5X29NXV0sIFN5c3RlbS5Qcml2YXRlLkNvcmVMaWIiLA0KICAgICAgIiR2YWx1ZXMiOiBbDQogICAgICAgIHsNCiAgICAgICAgICAiJHR5cGUiOiAiQkgub00uR2VvbWV0cnkuUG9pbnQsIEdlb21ldHJ5X29NIiwNCiAgICAgICAgICAiWCI6IDAuMCwNCiAgICAgICAgICAiWSI6IDAuMCwNCiAgICAgICAgICAiWiI6IDAuMA0KICAgICAgICB9LA0KICAgICAgICB7DQogICAgICAgICAgIiR0eXBlIjogIkJILm9NLkdlb21ldHJ5LlBvaW50LCBHZW9tZXRyeV9vTSIsDQogICAgICAgICAgIlgiOiA1LjAsDQogICAgICAgICAgIlkiOiAwLjAsDQogICAgICAgICAgIloiOiAwLjANCiAgICAgICAgfSwNCiAgICAgICAgew0KICAgICAgICAgICIkdHlwZSI6ICJCSC5vTS5HZW9tZXRyeS5Qb2ludCwgR2VvbWV0cnlfb00iLA0KICAgICAgICAgICJYIjogOTkuMCwNCiAgICAgICAgICAiWSI6IDAuMCwNCiAgICAgICAgICAiWiI6IDAuMA0KICAgICAgICB9DQogICAgICBdDQogICAgfQ0KICB9DQp9"^^:BH.oM.Adapters.RDF.Base64JsonSerialized;
		:BH.oM.Architecture.Elements.Room.Location "ew0KICAiJHR5cGUiOiAiQkguRW5naW5lLkFkYXB0ZXJzLlJERi5Db252ZXJ0K1R5cGVXcmFwcGVyYDFbW0JILm9NLkdlb21ldHJ5LlBvaW50LCBHZW9tZXRyeV9vTV1dLCBSREZfRW5naW5lIiwNCiAgIlZhbHVlIjogew0KICAgICIkdHlwZSI6ICJCSC5vTS5HZW9tZXRyeS5Qb2ludCwgR2VvbWV0cnlfb00iLA0KICAgICJYIjogMC4wLA0KICAgICJZIjogMC4wLA0KICAgICJaIjogMC4wDQogIH0NCn0="^^:BH.oM.Adapters.RDF.Base64JsonSerialized;
		:BH.oM.Base.BHoMObject.BHoM_Guid "de0774d0-f338-4fcf-9c9c-4c8b44ed3a29"^^xsd:string;
		:BH.oM.Base.BHoMObject.Name "first room object"^^xsd:string .
```
You can see how the address at `<http://individuals.address/10b49b3d-f976-bb2a-4a04-2a7dc83a56b8>` does not contain the Room Guid, that is listed below as `:BH.oM.Base.BHoMObject.BHoM_Guid "de0774d0-f338-4fcf-9c9c-4c8b44ed3a29"^^xsd:string;`. I think we want the address to be pointing at the correct individual ID, which for Room is the BHoM_Guid, whilst for Geometry is the Guid of the hash.

You tests failed to capture this because you were looking for the presence of the Guid or Guid of the Hash within the exported TTL, but: 
- the BHoM_Guid will always be present (as a property of the object);
- for IObjects, because the ontology identifier is never a property of the object, the check you've written should still be safe (in fact it does pass even if I change the check to my proposed version), but it is safer to check for the full URI (as it is the intended target of the identifier - being part of the URI).

For this reason, I think we should modify the check as I've proposed (will do a commit myself on this, I've added a Nuget package to make it simpler). Haven't had time to put thought on what else we need to do, but feel free to have another look or we can continue together.

File: RDF_Tests/ToTTLTests.cs (hunk 1 above)
Comment (alelom) [reply RDF_Tests/ToTTLTests.cs]: Done in https://github.com/BHoM/RDF_Prototypes/pull/68/commits/c0f70ecdd9093e071ee5bc2837888cd65306d24e

File: RDF_Tests/ToTTLTests.cs (hunk 2 above)
Comment (alelom) [reply RDF_Tests/ToTTLTests.cs]: Done in https://github.com/BHoM/RDF_Prototypes/pull/68/commits/c0f70ecdd9093e071ee5bc2837888cd65306d24e

**Review by DiellzaElshani — APPROVED** [review]
image
 works well on my end.

**Review by alelom — APPROVED** [review]

**Comment by alelom** [comment]
I'm merging this bypassing branch protection as adding @polnischfrosch to the BHoM org still didn't solve the problem unfortunately. We will have to look into that.
image

**Pull request merged**